### PR TITLE
RDISCROWD-893 Chat integration

### PR DESCRIPTION
**Pull request by kbecker42**
- Added chat integration with Slack and Chatlio.

## Enable Chatlio and Slack integration

For a specific list of projects by `short_name` or `all`.

```
CHAT_ENABLE = ['all']
CHAT_URL = 'https://gigwork-dev.slack.com'
```

```
CHAT_ENABLE = ['project1', 'project2']
CHAT_URL = 'https://gigwork-dev.slack.com'
```

Related https://github.com/bloomberg/pybossa-default-theme/pull/218

### CODE REVIEW

File: pybossa/api/__init__.py [hunk 1]
```diff
@@ -376,6 +379,57 @@ def get_disqus_sso_api():
...
+    try:
+        data = request.json
+        subject = 'Chat session started for project ' + short_name + ' by ' + current_user.email_addr
+        success_body = (
+            'A user has started a chat session on a project that you an owner/co-owner for.\n\n'
+            '    Project Short Name: {short_name}\n'
+            '    User requesting assistance: {user}\n'
+            '    Message: {message}\n\n'
+            'Slack Url\n'
+            'https://gigwork-dev.slack.com\n'
```
Comment (fiorda) [inline pybossa/api/__init__.py]: can we have a config parameter for this?

File: pybossa/api/__init__.py [hunk 2]
```diff
@@ -376,6 +379,57 @@ def get_disqus_sso_api():
         return error.format_exception(e, target='DISQUS_SSO', action='GET')
 
 
+@jsonpify
+@csrf.exempt
```
Comment (fiorda) [inline pybossa/api/__init__.py]: why exempt?

File: pybossa/api/__init__.py [hunk 3]
```diff
@@ -376,6 +379,57 @@ def get_disqus_sso_api():
...
+            'https://gigwork-dev.slack.com\n'
+            )
+
+        body = success_body.format(
+            short_name=project.short_name,
+            user=current_user.email_addr,
+            message=data.get('message'))
+
+        # Get email addresses for all owners of the project.
+        recipients = []
```
Comment (fiorda) [inline pybossa/api/__init__.py]:
```python
recipients = [user.email_addr for user in user_rep.get_users(project.owners_ids)]
```

File: pybossa/api/__init__.py [hunk 4]
```diff
@@ -376,6 +379,57 @@ def get_disqus_sso_api():
...
+        for user in user_repo.get_users(project.owners_ids):
+            recipients.append(user.email_addr)
+        import pdb
```
Comment (fiorda) [inline pybossa/api/__init__.py]: this might need to be removed

File: pybossa/api/__init__.py [hunk 5]
```diff
@@ -376,6 +379,57 @@ def get_disqus_sso_api():
...
+        pdb.set_trace()
+        email = dict(recipients=recipients,
+                     subject=subject,
+                     body=body)
+        mail_queue.enqueue(send_mail, email)
+
+        return Response(json.dumps({'success': True}), 200, mimetype="application/json")
```
Comment (fiorda) [inline pybossa/api/__init__.py]: also see https://flask.palletsprojects.com/en/1.1.x/api/#flask.json.jsonify

File: pybossa/api/__init__.py (hunk 1 above)
Comment (kbecker42) [reply pybossa/api/__init__.py]: Added via https://github.com/bloomberg/pybossa/pull/400/commits/5be401fc9581b2ccfcd416b6c907b8e7f4cca3e7

File: pybossa/api/__init__.py (hunk 2 above)
Comment (kbecker42) [reply pybossa/api/__init__.py]: This appears to be required for the ajax call to succeed from the client.

File: pybossa/api/__init__.py (hunk 3 above)
Comment (kbecker42) [reply pybossa/api/__init__.py]: Fixed via https://github.com/bloomberg/pybossa/pull/400/commits/5be401fc9581b2ccfcd416b6c907b8e7f4cca3e7

File: pybossa/api/__init__.py (hunk 4 above)
Comment (kbecker42) [reply pybossa/api/__init__.py]: :)
https://github.com/bloomberg/pybossa/pull/400/commits/5be401fc9581b2ccfcd416b6c907b8e7f4cca3e7

File: pybossa/api/__init__.py (hunk 2 above)
Comment (fiorda) [reply pybossa/api/__init__.py]: You should pass a CSRF token with your POST request

Comment (kbecker42) [reply pybossa/api/__init__.py]: Fixed via https://github.com/bloomberg/pybossa/pull/400/commits/0db33773e60b99354a9fdea652b40bbe1330de8e

**Pull request merged**